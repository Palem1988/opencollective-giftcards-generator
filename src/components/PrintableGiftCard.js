import React from 'react';
import PropTypes from 'prop-types';
import { Flex, Box } from '@rebass/grid';
import { FormattedMessage, FormattedDate } from 'react-intl';
import QRCode from 'qrcode.react';
import styled from 'styled-components';
import { borderRadius, fontSize } from 'styled-system';

import { ExternalLink } from 'styled-icons/feather/ExternalLink';

import { imgUrl } from '../lib/utils';
import { WEBSITE_URL } from '../constants/env';
import Container from './Container';
import StyledHr from './StyledHr';
import { P, Span } from './Text';
import Currency from './Currency';

const Card = styled(Box)`
  font-family: Helvetica, sans-serif;
  width: 3.5in;
  height: 2in;
  position: relative;
  overflow: hidden;
  background-image: url('${imgUrl('oc-gift-card-front-straightened.png')}');
  background-size: 100%;
  background-repeat: no-repeat;
  border: 1px solid lightgrey;
  color: white;
  display: flex;
  flex-direction: column;
  justify-content: space-between;

  @media print {
    break-inside: avoid;
  }

  ${borderRadius};
  ${fontSize};
`;

const OpenCollectiveLogo = styled.img`
  width: 3em;
  height: 3em;
`;

/**
 * A static gift card meant to be printed to be offered to someone. It has standard
 * business card resolution (3.5in x 2in) but is rendered two time bigger for HDPI.
 */
const PrintableGiftCard = ({ amount, currency, code, expiryDate, tagline, withQRCode, ...styleProps }) => {
  const shortWebsiteUrl = WEBSITE_URL.replace('https://', '');
  const redeemUrl = `${WEBSITE_URL}/redeem/${code}`;
  const basePaddingX = '0.8em';
  const paddingTop = '1em';

  return (
    <Card {...styleProps}>
      {/** Header */}
      <Flex justifyContent="space-between" alignItems="center" px={basePaddingX} pt={paddingTop}>
        <Flex alignItems="center">
          <OpenCollectiveLogo src={imgUrl('opencollective-icon.svg')} alt="" />
          <Flex flexDirection="column" ml="0.8em">
            <P fontWeight="bold" fontSize="1.1em" lineHeight="1.5em">
              Open Collective
            </P>
            {tagline && (
              <P fontSize="0.7em" lineHeight="0.8em" color="white.transparent.72">
                {tagline}
              </P>
            )}
          </Flex>
        </Flex>

        <Container
          background="#69a0f1"
          color="#d7e8fe"
          borderRadius="1em"
          padding="0.25em 1em"
          boxShadow="2px 3px 5px rgba(0, 0, 0, 0.15)"
          fontWeight="bold"
          fontSize="0.7em"
        >
          <FormattedMessage id="GiftCard" defaultMessage="Gift card" />
        </Container>
      </Flex>

      {/** Content */}
      <Flex justifyContent="space-between" flex="1 1" mb="0.3em">
        {/** Left */}
        <Flex flexDirection="column" justifyContent="space-between" ml={basePaddingX}>
          <Box>
            <StyledHr mt="0.75em" mb="0.5em" borderColor="rgb(62, 129, 230)" borderRadius={8} />
            {expiryDate && (
              <P fontSize="0.55em" lineHeight="1.75em" color="black.300">
                <FormattedMessage
                  id="ContributePayment.expiresOn"
                  defaultMessage="Expires on {expiryDate}"
                  values={{
                    expiryDate: <FormattedDate value={expiryDate} day="numeric" year="numeric" month="long" />,
                  }}
                />
              </P>
            )}
          </Box>
          <P fontSize="0.8em">
            <ExternalLink size="1em" color="black" />
            <Span color="black.500" ml={1}>
              {shortWebsiteUrl}/redeem/
            </Span>
            <Span fontWeight="bold" color="black.800">
              {code}
            </Span>
          </P>
        </Flex>

        {/** Right */}
        <Flex flexDirection="column" justifyContent="flex-end" alignItems="flex-end" pr={basePaddingX}>
          <Flex flexDirection="column" justifyContent="flex-end" alignItems="flex-end">
            {withQRCode && (
              <Container height="4.5em" border="1px solid lightgrey" boxShadow="2px 5px 4px rgba(0,0,0,0.15)">
                {/** Use 2x the real size to get a better resolution */}
                <QRCode
                  renderAs="svg"
                  value={redeemUrl}
                  size={256}
                  fgColor="#313233"
                  style={{ width: '4.5em', height: '4.5em' }}
                />
              </Container>
            )}
            <Flex mt="1em" mb="0.1em">
              <Span fontWeight="bold" fontSize="2em" lineHeight="1em" color="#313233">
                <Currency value={amount} currency={currency} precision={0} />
              </Span>
              <Box ml="0.25em">
                <Span color="black.700" fontSize="0.9em" className="currency">
                  {currency}
                </Span>
              </Box>
            </Flex>
          </Flex>
        </Flex>
      </Flex>
    </Card>
  );
};

PrintableGiftCard.propTypes = {
  /** The amount in cents */
  amount: PropTypes.number.isRequired,
  /** Currency of the gift card (eg. `EUR`) */
  currency: PropTypes.string.isRequired,
  /** The 8 characters code of the gift card */
  code: PropTypes.string.isRequired,
  /** Expiry date */
  expiryDate: PropTypes.string,
  /** Border radius for the card */
  borderRadius: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /** If true, QR code will be displayed */
  withQRCode: PropTypes.bool,
  /** The tagline displayed under Open Collective logo */
  tagline: PropTypes.node,
  /** Main font size */
  fontSize: PropTypes.string,
};

PrintableGiftCard.defaultProps = {
  tagline: 'Transparent funding for open source',
  withQRCode: true,
  borderRadius: '0px',
  fontSize: '0.14in',
};

export default PrintableGiftCard;
