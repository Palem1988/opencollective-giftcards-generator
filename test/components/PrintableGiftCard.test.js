import React from 'react';
import renderer from 'react-test-renderer';
import { IntlProvider } from 'react-intl';

import PrintableGiftCard from '../../src/components/PrintableGiftCard';

it('renders correctly', () => {
  const tree = renderer
    .create(
      <IntlProvider locale="en">
        <PrintableGiftCard
          amount={50000}
          currency="USD"
          code="8X4WWD2G"
          expiryDate="Wed May 13 2020 00:00:00 GMT+0200 (GMT+02:00)"
          withQRCode={false}
          tagline="I have no QR code!"
          style={{ marginRight: 15 }}
        />
      </IntlProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
