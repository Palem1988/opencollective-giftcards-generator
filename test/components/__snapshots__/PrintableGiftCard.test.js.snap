// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c10 {
  box-sizing: border-box;
}

.c23 {
  box-sizing: border-box;
  margin-left: 0.25em;
}

.c1 {
  box-sizing: border-box;
  padding-top: 1em;
  padding-left: 0.8em;
  padding-right: 0.8em;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c2 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 {
  box-sizing: border-box;
  margin-left: 0.8em;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c8 {
  box-sizing: border-box;
  margin-bottom: 0.5em;
  -webkit-flex: 1 1;
  -ms-flex: 1 1;
  flex: 1 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c9 {
  box-sizing: border-box;
  margin-left: 0.8em;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c18 {
  box-sizing: border-box;
  padding-right: 0.8em;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c19 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c20 {
  box-sizing: border-box;
  margin-top: 1em;
  margin-bottom: 0.1em;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c15 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c7 {
  box-sizing: border-box;
  background: #69a0f1;
  border-radius: 1em;
  border-radius: 1em;
  box-shadow: 2px 3px 5px rgba(0,0,0,0.15);
  color: #d7e8fe;
  font-weight: bold;
  font-size: 0.75em;
  padding: 0.25em 1em;
}

.c11 {
  border: 0;
  border-top: 1px solid;
  margin: 0;
  height: 1px;
  margin-top: 0.75em;
  margin-bottom: 0.5em;
  border-color: rgb(62,129,230);
}

.c5 {
  font-size: 1.1em;
  font-weight: bold;
  line-height: 1.5em;
  -webkit-letter-spacing: -0.2px;
  -moz-letter-spacing: -0.2px;
  -ms-letter-spacing: -0.2px;
  letter-spacing: -0.2px;
  margin: 0;
}

.c6 {
  color: white.transparent.72;
  font-size: 0.75em;
  line-height: 0.8em;
  -webkit-letter-spacing: -0.2px;
  -moz-letter-spacing: -0.2px;
  -ms-letter-spacing: -0.2px;
  letter-spacing: -0.2px;
  margin: 0;
}

.c12 {
  color: black.100;
  font-size: 0.6em;
  line-height: 1.75em;
  -webkit-letter-spacing: -0.2px;
  -moz-letter-spacing: -0.2px;
  -ms-letter-spacing: -0.2px;
  letter-spacing: -0.2px;
  margin: 0;
}

.c13 {
  color: black.300;
  font-size: 0.55em;
  line-height: 1.75em;
  -webkit-letter-spacing: -0.2px;
  -moz-letter-spacing: -0.2px;
  -ms-letter-spacing: -0.2px;
  letter-spacing: -0.2px;
  margin: 0;
}

.c14 {
  font-size: 0.8em;
  line-height: Paragraph;
  -webkit-letter-spacing: -0.2px;
  -moz-letter-spacing: -0.2px;
  -ms-letter-spacing: -0.2px;
  letter-spacing: -0.2px;
  margin: 0;
  margin-top: 0.05in;
  margin-bottom: 0.05in;
}

.c16 {
  color: black.500;
  font-size: inherit;
  line-height: inherit;
  -webkit-letter-spacing: -0.2px;
  -moz-letter-spacing: -0.2px;
  -ms-letter-spacing: -0.2px;
  letter-spacing: -0.2px;
  margin: 0;
  margin-left: 4px;
}

.c17 {
  color: black.800;
  font-size: inherit;
  font-weight: bold;
  line-height: inherit;
  -webkit-letter-spacing: -0.2px;
  -moz-letter-spacing: -0.2px;
  -ms-letter-spacing: -0.2px;
  letter-spacing: -0.2px;
  margin: 0;
}

.c21 {
  color: #313233;
  font-size: 2em;
  font-weight: bold;
  line-height: 1em;
  -webkit-letter-spacing: -0.2px;
  -moz-letter-spacing: -0.2px;
  -ms-letter-spacing: -0.2px;
  letter-spacing: -0.2px;
  margin: 0;
}

.c22 {
  font-size: inherit;
  line-height: inherit;
  -webkit-letter-spacing: -0.2px;
  -moz-letter-spacing: -0.2px;
  -ms-letter-spacing: -0.2px;
  letter-spacing: -0.2px;
  margin: 0;
}

.c24 {
  color: black.700;
  font-size: 0.9em;
  line-height: inherit;
  -webkit-letter-spacing: -0.2px;
  -moz-letter-spacing: -0.2px;
  -ms-letter-spacing: -0.2px;
  letter-spacing: -0.2px;
  margin: 0;
}

.c0 {
  box-sizing: border-box;
  width: 7in;
  height: 4in;
  position: relative;
  overflow: hidden;
  background-image: url('/static/images/oc-gift-card-front-straightened.svg');
  background-size: 100%;
  background-repeat: no-repeat;
  border: 1px solid lightgrey;
  color: white;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  font-size: 0.28in;
  border-radius: 0px;
}

.c3 {
  width: 3em;
  height: 3em;
}

@media print {
  .c0 {
    width: 3.5in;
    height: 2in;
    font-size: 0.14in;
    -webkit-break-inside: avoid;
    break-inside: avoid;
  }
}

<div
  className="c0"
  style={
    Object {
      "marginRight": 15,
    }
  }
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <img
        alt=""
        className="c3"
        src="/static/images/opencollective-icon.svg"
      />
      <div
        className="c4"
      >
        <p
          className="c5"
        >
          Open Collective
        </p>
        <p
          className="c6"
        >
          I have no QR code!
        </p>
      </div>
    </div>
    <div
      className="c7"
      padding="0.25em 1em"
    >
      <span>
        Gift card
      </span>
    </div>
  </div>
  <div
    className="c8"
  >
    <div
      className="c9"
    >
      <div
        className="c10"
      >
        <hr
          className="c11"
        />
        <p
          className="c12"
        >
          Contribute to the project of your choice at 
          <strong>
            opencollective.com
          </strong>
        </p>
        <p
          className="c13"
        >
          <span>
            Expires on 
            <span>
              May 12, 2020
            </span>
          </span>
        </p>
      </div>
      <p
        className="c14"
      >
        <svg
          aria-hidden="true"
          className="c15"
          color="black"
          fill="none"
          focusable="false"
          height="1em"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          viewBox="0 0 24 24"
          width="1em"
        >
          <path
            d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
          />
          <polyline
            points="15 3 21 3 21 9"
          />
          <line
            x1={10}
            x2={21}
            y1={14}
            y2={3}
          />
        </svg>
        <span
          className="-Clean-span c16"
        >
          opencollective.com
          /redeem/
        </span>
        <span
          className="-Clean-span c17"
        >
          8X4WWD2G
        </span>
      </p>
    </div>
    <div
      className="c18"
    >
      <div
        className="c19"
      >
        <div
          className="c20"
        >
          <span
            className="-Clean-span c21"
          >
            <span
              className="-Clean-span c22"
            >
              $500
            </span>
          </span>
          <div
            className="c23"
          >
            <span
              className="currency -Clean-span c24"
            >
              USD
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
